File: 4020-lexicographically-smallest-permutation-greater-than-target/lexicographically-smallest-permutation-greater-than-target.java
class Solution {
    public String lexGreaterPermutation(String s, String target) {
        int n = s.length();
        int[] count = new int[26];
        for (char c : s.toCharArray()) {
            count[c - 'a']++;
        }

        int[] prefixCount = count.clone();
        int matchLen = 0;
        while (matchLen < n && prefixCount[target.charAt(matchLen) - 'a'] > 0) {
            prefixCount[target.charAt(matchLen) - 'a']--;
            matchLen++;
        }

        for (int i = matchLen; i >= 0; i--) {
            if (i < matchLen) {
                prefixCount[target.charAt(i) - 'a']++;
            }
            if (i == n) continue;

            int targetChar = target.charAt(i) - 'a';
            for (int c = targetChar + 1; c < 26; c++) {
                if (prefixCount[c] > 0) {
                    StringBuilder sb = new StringBuilder();
                    sb.append(target, 0, i);
                    sb.append((char) ('a' + c));
                    prefixCount[c]--;

                    for (int ch = 0; ch < 26; ch++) {
                        while (prefixCount[ch] > 0) {
                            sb.append((char) ('a' + ch));
                            prefixCount[ch]--;
                        }
                    }
                    return sb.toString();
                }
            }
        }

        return "";
    }
}

File: 4020-lexicographically-smallest-permutation-greater-than-target/lexicographically-smallest-permutation-greater-than-target.py
class Solution:
    def lexGreaterPermutation(self, s: str, target: str) -> str:
        n = len(s)
        count = [0] * 26
        for c in s:
            count[ord(c) - 97] += 1
        
        prefix_count = count.copy()
        match_len = 0
        while match_len < n and prefix_count[ord(target[match_len]) - 97] > 0:
            prefix_count[ord(target[match_len]) - 97] -= 1
            match_len += 1
        
        for i in range(match_len, -1, -1):
            if i < match_len:
                prefix_count[ord(target[i]) - 97] += 1
            if i == n:
                continue
            
            target_char = ord(target[i]) - 97
            for c in range(target_char + 1, 26):
                if prefix_count[c] > 0:
                    res = [target[:i], chr(97 + c)]
                    prefix_count[c] -= 1
                    for ch in range(26):
                        if prefix_count[ch] > 0:
                            res.append(chr(97 + ch) * prefix_count[ch])
                    return "".join(res)
        
        return ""